draft-ietf-protected-audience-key-value-services.md

```diff
@@ -0,0 +1,717 @@
+following schema for the client to
+successfully parse them.
+
+Note that they must be serialized to string when stored as the value.
+
+#### InterestGroupResponse {#interest-group-response}
+
+The schema below is defined following the spec by
+https://json-schema.org. For values for keys
+from the `interestGroupNames` namespace, they must conform to the
+following schema, prior to being
+serialized to string.
+
+~~~~~ json
+{
+    "title": "tkv.response.v2.InterestGroupResponse",
+    "description": "Format for value of keys in groups tagged 'interestGroupNames'",
+    "type": "object",
+    "additionalProperties": false,
+    "properties": {
+        "priorityVector": {
+            "type": "object",
+            "patternProperties": {
+                ".*": {
+                    "description": "signals",
+                    "type": "number"
+                }
+            }
+        },
+        "updateIfOlderThanMs": {
+            "description": "This optional field specifies that the interest group should be updated if the interest group hasn't been joined or updated in a duration of time exceeding `updateIfOlderThanMs` milliseconds. Updates that ended in failure, either parse or network failure, are not considered to increment the last update or join time. An `updateIfOlderThanMs` that's less than 10 minutes will be clamped to 10 minutes.",
+            "type": "unsigned integer"
+        }
+    }
+}
+~~~~~
+
+### Generating a Response
+
+This algorithm describes how the Key Value Service MAY generate a response to a request.
+
+The input is a list of [deterministically encoded CBOR](https://www.rfc-editor.org/rfc/rfc8949.html#name-deterministically-encoded-c) `partitionOutputs` in {{response-schema}} as well as
+the `compression group map` and the HPKE receiver, `rctxt`, context saved in {{request-parsing}}.
+
+The output is a `response` to be sent to a Client.
+
+1. Create `payload`.
+2. Set `compression groups` to an empty list.
+3. Set `partition output map` to an empty map.
+4. For each `partitionOutput` in the list of `partitionOutputs`:
+   1. Set `partition id` to `partitionOutput["id"]` if present, otherwise 0.
+   2. Add key value pair `partition id`, `partitionOutput` to `partition output map`.
+5. For each (`compression group id`, `partition ids`) in `compression group map`:
+   1. Create `compression group`.
+   2. Set `cbor partitions array` to an empty CBOR array.
+   3. For each `partition id` in `partition ids`:
+      1. Set `partition output` to `partition output map[partition id]`.
+         1. On failure to find the partition, continue.
+      2. Add `partition output` to `cbor partitions array`.
+   4. Set `cbor serialized payload` to the CBOR serialized `cbor partitions array`.
+      1. On serialization failure, continue.
+   5. Set `compression group content` to the [GZIP] compressed `cbor serialized payload`.
+      1. On failure, set `compression group content` to empty.
```

Comment:
So on compression failure, we send an empty message field instead of omitting it entirely? It seems like we should just skip the entire compression group.

Reply:
Done.